tox_docker/tox4/config.py

```diff
@@ -0,0 +1,189 @@
+from collections import defaultdict
+from typing import Callable, Container, Dict, List, Sequence
+
+from tox.config.loader.section import Section
+from tox.config.main import Config
+from tox.config.sets import ConfigSet
+from tox.tox_env.api import ToxEnv
+
+from tox_docker.config import (
+    ContainerConfig,
+    RunningContainers,
+    validate_link,
+    validate_port,
+    validate_volume,
+)
+
+# nanoseconds in a second; named "SECOND" so that "1.5 * SECOND" makes sense
+SECOND = 1000000000
+
+EnvRunningContainers = Dict[ToxEnv, RunningContainers]
+
+
+class MissingRequiredSetting(Exception):
+    pass
+
+
+def required(setting_name: str) -> Callable[[str], str]:
+    def require_value(val: str) -> str:
+        if not val:
+            raise MissingRequiredSetting(setting_name)
+        return val
+
+    return require_value
+
+
+class EnvDockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["docker"],
+            of_type=List[str],
+            default=[],
+            desc="docker image configs to load",
+        )
+
+
+class DockerConfigSet(ConfigSet):
+    def register_config(self) -> None:
+        self.add_config(
+            keys=["image"],
+            of_type=str,
+            default="",
+            post_process=required("image"),
+            desc="docker image to run",
+        )
+        self.add_config(
+            keys=["environment"],
+            of_type=Dict[str, str],
+            default={},
+            desc="environment variables to pass to the docker container",
+        )
+        self.add_config(
+            keys=["ports"],
+            of_type=List[str],
```

Comment:
shouldn't be list of int this one?

Reply:
No, they're strings like "8080:80/tcp" potentially. Again, the restriction that the initially-parsed type and the postprocessed type must match prevents me from doing "the right thing" here (to do the final step of parsing as part of the config loading)

Reply:
I don't think you need that indirection. Why don't you create a custom type that takes the raw string as init argument and validates/builds it? E.g. introduce the DockerPort type 😊 Mark it as a type of that, the class takes a raw argument; which can be Any. If it's string parse/validate it; if it's of DockerType just clone it. Configuration values passed from `toxfile.py` may decide to pass in the raw type directly 🤔